Feature/notebook

Cyrik opened the pull request:
<!-- ❤️ Thanks for filing a Pull Request on Calva! You are contributing to a better Clojure coding experience. ❤️
## What you can expect:

Here are some things we consider before we merge:

- We make sure the PR is directed at the `dev` branch (unless reasons).
- We figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and will help you test there if it is hard for you to do so. (We appreciate a lot if you take on the work do this of course.)
- We read the source changes. (Surprise! 😄)
- We given feedback and guidance on source changes, if needed. Far from everything is captured in our [code guidelines](https://github.com/BetterThanTomorrow/calva/wiki/Coding-Style).
- We use our domain knowledge to try catch if you have missed some facility already provided in the code base.
- We read the updates to the documentation and help with feedback, trying to keep the documentation site serving well.
- We often check out your code changes and test them.
- We sometimes send the VSIX built from the PR out in the `#calva` channel on slack for others to test. (Actually, we will probably encourage you to do this.)
- We sometimes have a chat within the team about particular changes.
- NB: We also consider if your changes belong in the Calva product we want to maintain. Before you spend a lot of work on a PR, please consider chatting us up first, and filing issues.

We try to be speedy and attentive. Please don't hesitate to bump a PR, or contact us, if we seem to have dropped the ball (that has happened).

We use checklists in order to not forget about important lessons we and others have learnt along the way.

-->

## What has changed?

<!-- Introduce the change(s) briefly here. Consider explaining why a particular change was implemented the way it was. If you have considered alternative ways to introduce the change, please elaborate a bit on that as well. -->

- adds a notebook provider
- adds a Notebook Kernel
You can try it out by right clicking on any clojure file -> `Open with...` -> `Calva Notebook`.

<!-- Tell us what Github issue(s) your PR is fixing. Consider creating the issue if there isn't one already. -->
Closes #1824 

## My Calva PR Checklist
<!-- Strike out (using `~`) items that do not apply, as Github reports how many are not ticked. If you want to add checkboxes, please do. -->

If this PR involves code changes, I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [x] Tests
  - [x] Tested the particular change
  - [x] Figured if the change might have some side effects and tested those as well.
  - [x] Smoke tested the extension as such.
  - [x] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
  - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
  - [x] If I am fixing just part of the issue, I have just referenced it w/o any of the "fixes” keywords.
- [x] Created the issue I am fixing/addressing, if it was not present.
- [x] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).

<!-- This is a nice book to read about the power of checklists: https://www.samuelthomasdavies.com/book-summaries/health-fitness/the-checklist-manifesto/ -->

Ping @pez, @bpringe, @corasaurus-hex, @Cyrik

Cyrik committed:
notebook support

Cyrik committed:
added html and output and highliting

Cyrik committed:
added saving

Cyrik committed:
prettier json for some reason

Cyrik commented:
So this is working "fully".
There's only one thing I'd like to fix before releasing this as MVP: saving currently reformats your file a little since I don't keep track of the whitespace between code blocks. @PEZ I'm using the LispCursor to get the top-level forms, is there something better that would also give me the whitespace ranges? Otherwise, I'll build it myself.
Other than that I'm being very unsafe with error handling and pretty printing, but I think that's ok for now, especially since this is a different view on the file that you can ignore if it's not working for your particular output size.

The next steps would be to use the configured pretty printer and to get all the usual tooling running, like autocomplete and stuff. Only a tiny bit of that is actually something we'll have to do in Calva, most of it is going to happen in clojure-lsp.

Cyrik committed:
cleanup

PEZ commented:
Thanks for this work, @Cyrik! ❤️ 

I'm unfamiliar with notebooks. Can you explain a bit about what the provider and kernel responsibilities are?

> I'm using the LispCursor to get the top-level forms, is there something better that would also give me the whitespace ranges? Otherwise, I'll build it myself.

Not really. There are some facilities for collecting whitespace information around the cursor, but I don't know how much use you could have for those. If _whitespace_ here means _everything else, but top level forms_, i.e. including comments, then the whitespace ranges are something like the inverse of the top level ranges, I think. Otherwise... if comments should be editable, maybe we should have a version of the top-level-ranges that includes ranges for the comment blocks?

I'd like to merge this, as is, and then we can build on it. However, would be good with a docs page for Notebooks so that users can orient themselves a bit on what to expect.

Cyrik commented:
Sure, but I'll have to keep it brief since I have to run for now.
The notebookProvider is a class that's gets called by vscode when you open or save a notebook. The deserialize method gets the content of the opened file and is supposed to return notebook cells that represent the file. The serialize method gets called on save and gets the cells and need to transform them into a string that is going to be written to the file.
You can go more complicated here and load arbitrary stuff from other sources or do transformations, but my basic implementation just wants to make every top level sexp into its own code block. To perserve comments and whitespace I'll have to go a little extra so that I can write them back out on save.

The inverse sounds like an OK way to go for now, I'll try to add that this to the PR. I'll also add a little bit to the docs. As you said, we can make this better in follow-up PRs and this feature is opt-in anyway.

Cyrik committed:
fixed notebook representation of whitespace and comments

Cyrik committed:
docs

Cyrik commented:
Fixed saving of whitespace/comments.
Handling of this could be a lot better, but it will save out the same input, so I'm happy for an alpha.

Also added a doc, although I'm not sure what to put into it 😅 

Happy for any suggestions for v0, otherwise I'd add a changelog entry and call this done, and work on improvements in the next version.

bpringe commented on package-lock.json:
```diff
@@ -1294,6 +1294,7 @@
       "version": "1.1.0",
       "resolved": "https://registry.npmjs.org/@npmcli/fs/-/fs-1.1.0.tgz",
       "integrity": "sha512-VhP1qZLXcrXRIaPoqb4YA55JQxLNF3jNR4T55IdOJa3+IFJKNYHtPvtXx8slmeMavj37vCzCfrqQM1vWLsYKLA==",
+      "deprecated": "this version had an improper engines field added, update to 1.1.1",
```
yikes... I'll make a task for this

bpringe replied:
And the other "deprecated" dep issues in this file

bpringe replied:
https://github.com/BetterThanTomorrow/calva/issues/1822

bpringe commented on test-data/notebook.clj:
```diff
@@ -0,0 +1,15 @@
+
+hover-map
+
+"<html> <body>testing<h2>HTML Image</h2><img src='https://raw.githubusercontent.com/BetterThanTomorrow/calva/dev/assets/calva-64h.png'></body></html>"
```
Prime opportunity for a Rickroll.

Cyrik replied:
done

bpringe reviewed (changes requested):
Just need a changelog entry and I think we're good to merge. 🎉 Thanks again for this!

PEZ committed:
Add some more notebook testing

PEZ committed:
Update docs

PEZ commented:
Been playing a bit with this now and I must say it's lovely. ❤️ This could very well get me started as a notebooks user.

I've updated the documentation page a bit and added it to the **Features** section.

**NB**: There is a quite nasty bug with comments preceding the first form not being retained. 

Some suggestions:

* I think the name **Clojure Notebook** is better than **Calva Notebook**. I used this name in the docs update, but haven't changed it in the config/code.
* I'd like an issue created about adding basic Notebook support. It could include a bit about what notebooks are and why it would be good to have them. This issue is what should be referred in the change log.
* Related to the issue ticket, I think we should start a Discussions thread about Notebooks. It seems to be a very large subject with lots of choices and priorities and trade-offs.

Cyrik committed:
rename and fix initial whitespace

PEZ reviewed (commented):
See my comments above. =)

Cyrik committed:
updated changelog

Cyrik commented:
I've renamed the description to "Clojure Notebook" but set the id to "calva-clojure-notebook" to prevent clashes with a potential extension that also provides Clojure notebooks.
Issue and discussion are also created and the "comment or whitespace before the first exp" is also fixed.
I think this can be merged @PEZ

bpringe commented on docs/site/notebooks.md:
```diff
@@ -0,0 +1,19 @@
+
+   Please help test the feature. We're looking forward to your feedback!
+
+You can open any clojure file as a notebook by right clicking the file -> `Open with...` -> `Calva Notebook`.
```
```suggestion
You can open any Clojure file as a notebook by right clicking the file -> `Open with...` -> `Clojure Notebook`.
```

bpringe replied:
I think this should be updated, right?

bpringe reviewed (approved):
Approving, but I made a doc update suggestion.

Cyrik committed:
Update docs/site/notebooks.md

Co-authored-by: Brandon Ringe <brandon.ringe@gmail.com>
